airgen/src/lib.rs

```diff
@@ -0,0 +1,193 @@
+            (
+                // get the absolute name for this submachine
+                location.clone().join(def.name.clone()),
+                // get its type
+                def.ty.clone(),
+            )
+        }));
+
+        instances.push((location, ty));
+    }
+
+    // visit the tree compiling the machines
+    let objects = instances
+        .into_iter()
+        .map(|(location, ty)| {
+            let object = ASMPILConverter::convert_machine(&location, &ty, &input);
+            (location, object)
+        })
+        .collect();
+
+    let main_ty = input.machines.get(&main_ty).unwrap();
+
+    PILGraph {
+        main: ast::object::Machine {
+            location: main_location,
+            function_id: main_ty.function_id.clone(),
+            latch: main_ty.latch.clone(),
+        },
+        entry_points: main_ty
+            .functions
+            .iter()
+            .map(|f| Function {
+                name: "main".to_string(),
+                id: f.id.unwrap_or(T::from(0)),
+                params: f.params.clone(),
+            })
+            .collect(),
+        objects,
+    }
+}
+
+struct ASMPILConverter<'a, T> {
+    /// Location in the machine tree
+    location: &'a Location,
+    /// Machine types
+    machines: &'a BTreeMap<String, Machine<T>>,
+    pil: Vec<PilStatement<T>>,
+    submachines: Vec<SubmachineDeclaration>,
+}
+
+impl<'a, T: FieldElement> ASMPILConverter<'a, T> {
+    fn new(location: &'a Location, input: &'a AnalysisASMFile<T>) -> Self {
+        Self {
+            location,
+            machines: &input.machines,
+            pil: Default::default(),
+            submachines: Default::default(),
+        }
+    }
+
+    fn handle_inline_pil(&mut self, PilBlock { statements, .. }: PilBlock<T>) {
+        self.pil.extend(statements)
+    }
+
+    fn convert_machine(
+        location: &'a Location,
+        ty: &'a str,
+        input: &'a AnalysisASMFile<T>,
+    ) -> Object<T> {
+        Self::new(location, input).convert_machine_inner(ty)
+    }
+
+    fn convert_machine_inner(mut self, ty: &str) -> Object<T> {
+        let input = self.machines.get(ty).unwrap().clone();
+
+        let degree = input.degree.map(|s| T::from(s.degree.clone()).to_degree());
+
+        self.submachines = input.submachines;
+
+        for block in input.constraints {
+            self.handle_inline_pil(block);
+        }
+
+        let links = input
+            .instructions
+            .into_iter()
+            .filter_map(|instr| self.handle_instruction_def(instr))
+            .collect();
+
+        Object {
+            degree,
+            pil: self.pil,
+            links,
+        }
+    }
+
+    fn handle_instruction_def(
+        &mut self,
+        InstructionDefinitionStatement {
+            start: _,
+            name,
+            params,
+            body,
+        }: InstructionDefinitionStatement<T>,
+    ) -> Option<Link<T>> {
+        let instruction_flag = format!("instr_{name}");
```

Comment:
Just to confirm, this doesn't require that the input has this column, but rather generates it here, right?

Reply:
will add todo, it will be fixed once we have CM links